src/approaches/pg3_approach.py

```diff
@@ -46,14 +47,39 @@ def __init__(self, initial_predicates: Set[Predicate],
     def get_name(cls) -> str:
         return "pg3"
 
-    def _predict(self, state: State, atoms: Set[GroundAtom],
-                 goal: Set[GroundAtom]) -> _GroundNSRT:
-        del state  # unused
+    def _predict_ground_nsrt(self, atoms: Set[GroundAtom],
+                             goal: Set[GroundAtom]) -> _GroundNSRT:
+        """Predicts next GroundNSRT to be deployed based on the PG3 generated
+        policy."""
         ground_nsrt = utils.query_ldl(self._current_ldl, atoms, goal)
         if ground_nsrt is None:
             raise ApproachFailure("PG3 policy was not applicable!")
         return ground_nsrt
 
+    def _solve(self, task: Task, timeout: int) -> Callable[[State], Action]:
+        """Searches for a low level policy that satisfies PG3's abstract
+        policy."""
+        skeleton = []
+        atoms_sequence = []
+        atoms = utils.abstract(task.init, self._initial_predicates)
+        atoms_sequence.append(atoms)
+        start_time = time.time()
+
+        while not task.goal.issubset(atoms):
+            if (time.time() - start_time) >= timeout:
+                raise ApproachFailure("Timeout exceeded")
+            ground_nsrt = self._predict_ground_nsrt(atoms, task.goal)
+            atoms = utils.apply_operator(ground_nsrt, atoms)
+            skeleton.append(ground_nsrt)
+            atoms_sequence.append(atoms)
+        option_list, succeeded = run_low_level_search(
+            task, self._option_model, skeleton, atoms_sequence, self._seed,
+            timeout - (time.time() - start_time), CFG.horizon)
+        if not succeeded:
+            raise ApproachFailure("Low Level Search Failed")
```

Comment:
minor: change capitalization to be consistent with elsewhere -- so, "Low-level search failed"